File: Code/Python/Auxiliary.py
#!/usr/bin/python
#  _     _                           ____  _       _         ____                            _ _   _
# | |   (_) ___ ___ _ __  ___  ___  |  _ \| | __ _| |_ ___  |  _ \ ___  ___ ___   __ _ _ __ (_) |_(_) ___  _ __
# | |   | |/ __/ _ \ '_ \/ __|/ _ \ | |_) | |/ _` | __/ _ \ | |_) / _ \/ __/ _ \ / _` | '_ \| | __| |/ _ \| '_ \
# | |___| | (__  __/ | | \__ \  __/ |  __/| | (_| | |_  __/ |  _ <  __/ (__ (_) | (_| | | | | | |_| | (_) | | | |
# |_____|_|\___\___|_| |_|___/\___| |_|   |_|\__,_|\__\___| |_| \_\___|\___\___/ \__, |_| |_|_|\__|_|\___/|_| |_|
#                                                                                |___/
# (c) Shahar Gino, July-2017, sgino209@gmail.com

from random import randint
from numpy import linspace, dstack
from onvif_capture import onvif_camera
from cv2 import resize, imshow, line, boxPoints, putText, getTextSize, FONT_HERSHEY_SIMPLEX, getRectSubPix,\
    COLOR_BGR2GRAY, cvtColor, Canny, matchTemplate, minMaxLoc, rectangle, TM_CCORR_NORMED, imread, imwrite

# ---------------------------------------------------------------------------------------------------------------
# Python structuring way:
class Struct:
    def __init__(self, **kwds):
        self.__dict__.update(kwds)

# ---------------------------------------------------------------------------------------------------------------
# Color vectors:
Colors = Struct(
    white=(255.0, 255.0, 255.0),
    green=(0.0, 255.0, 0.0),
    red=(0.0, 0.0, 255.0),
    yellow=(0.0, 255.0, 255.0)
)

# ---------------------------------------------------------------------------------------------------------------
def imshow_scaled(figure, img, dimensions):
    """ downscale + imshow """
    if img.shape[0] > dimensions[0] or img.shape[0] > dimensions[1]:
        img = resize(img, (img.shape[1] / 2, img.shape[0] / 2))
    imshow(figure, img)


# ---------------------------------------------------------------------------------------------------------------
def debug(message, debugMode):
    """ Auxiliary function for a conditional debug printout """

    if debugMode:
        print "DEBUG: %s" % message

# ---------------------------------------------------------------------------------------------------------------
def info(message):
    """ Auxiliary function for a info printout """

    print "INFO: %s" % message

# ---------------------------------------------------------------------------------------------------------------
def error(message):
    """ Auxiliary function for a error printout """

    print "ERROR: %s" % message

# ---------------------------------------------------------------------------------------------------------------
def drawRedRectangleAroundPlate(imgOriginalScene, licPlate, color):
    """ Mark a given license-plate (licPlate) with a colored rectangle """

    # Get 4 vertices of rotated rectangle:
    if licPlate.rrLocationOfPlateInSceneGbl is not None:
        p2fRectPoints = boxPoints(licPlate.rrLocationOfPlateInSceneGbl)
    else:
        p2fRectPoints = boxPoints(licPlate.rrLocationOfPlateInScene)

    # Draw 4 red lines:
    line(imgOriginalScene, tuple(p2fRectPoints[0]), tuple(p2fRectPoints[1]), color, 2)
    line(imgOriginalScene, tuple(p2fRectPoints[1]), tuple(p2fRectPoints[2]), color, 2)
    line(imgOriginalScene, tuple(p2fRectPoints[2]), tuple(p2fRectPoints[3]), color, 2)
    line(imgOriginalScene, tuple(p2fRectPoints[3]), tuple(p2fRectPoints[0]), color, 2)

# ---------------------------------------------------------------------------------------------------------------
def writeLicensePlateCharsOnImage(imgOriginalScene, licPlate, color):
    """ Write the given license-plate characters (licPlate.strChars) the provided image (imgOriginalScene) """

    # Retrieve basic geometric metrics:
    sceneHeight = imgOriginalScene.shape[0]
    plateHeight = licPlate.imgPlate.shape[0]

    # Font styling (type, scale, thickness):
    intFontFace = FONT_HERSHEY_SIMPLEX
    fltFontScale = float(plateHeight) / 30.0
    intFontThickness = int(round(fltFontScale * 1.5))
    textSize = getTextSize(licPlate.strChars, intFontFace, fltFontScale, intFontThickness)[0]

    # Retrieve license-plate centroid:
    if licPlate.rrLocationOfPlateInSceneGbl is not None:
        ((intPlateCenterX, intPlateCenterY), (_, _), _) = licPlate.rrLocationOfPlateInSceneGbl
    else:
        ((intPlateCenterX, intPlateCenterY), (_, _), _) = licPlate.rrLocationOfPlateInScene

    # The horizontal location of the text area is the same as the plate:
    ptCenterOfTextAreaX = int(intPlateCenterX)

    # The vertical location of the test area depends on the license-plate location in the scene image::
    if intPlateCenterY < (sceneHeight * 0.75):
        ptCenterOfTextAreaY = int(round(intPlateCenterY)) + int(round(plateHeight * 1.6))  # chars below the plate
    else:
        ptCenterOfTextAreaY = int(round(intPlateCenterY)) - int(round(plateHeight * 1.6))  # chars above the plate

    # Calculate the lower left origin of the text area based on the text area center, width, and height:
    ptLowerLeftTextOriginX = int(ptCenterOfTextAreaX - (textSize[0] / 2))
    ptLowerLeftTextOriginY = int(ptCenterOfTextAreaY + (textSize[1] / 2))

    # Write the text on the image:
    putText(imgOriginalScene, licPlate.strChars, (ptLowerLeftTextOriginX, ptLowerLeftTextOriginY), intFontFace, fltFontScale, color, intFontThickness)

# ---------------------------------------------------------------------------------------------------------------
def ROI_adjust(ROI, imgOriginalScene, W, H):
    """ ROI adjustments """

    imgH, imgW, _ = imgOriginalScene.shape

    if len(ROI) == 2 and ROI[0] == -1 and ROI[1] == -1:

        adjROI = (int(imgW * W[0]/100.0), int(imgH * H[0]/100.0),  int(imgW * W[1]/100.0), int(imgH * H[1]/100.0))
        autoRoiMode = True

        if sum(W) != 100 or sum(H) != 100:
            error('Invalid autoROI settings, doesnt converge to 100 percent: W=%s, H=%s' % (W, H))

    elif len(ROI) == 2 and ROI[0] == 0 and ROI[1] == 0:

        adjROI = (0, 0,  imgW, imgH)
        autoRoiMode = False

    else:
        adjROI = ROI
        autoRoiMode = False

    return adjROI, autoRoiMode

# ---------------------------------------------------------------------------------------------------------------
def crop_roi_from_image(imgOriginalScene, ROI, autoRoiMode):
    """ Crop a given ROI from a given Image; ROI=(startX,startY,W,H) """

    if ROI[2] == -1 or ROI[3] == -1:
        imgH, imgW, _ = imgOriginalScene.shape
        roiW = imgW
        roiH = imgH
    else:
        roiW = ROI[2]
        roiH = ROI[3]

    roiCx = ROI[0] + roiW / 2.0
    roiCy = ROI[1] + roiH / 2.0

    imgCropped = getRectSubPix(imgOriginalScene, (roiW, roiH), (roiCx, roiCy))

    info("ROI size: (Cx,Cy)=(%.2f,%.2f), WxH=%dx%d, autoROI=%d" % (roiCx, roiCy, roiW, roiH, int(autoRoiMode)))

    return imgCropped

# ---------------------------------------------------------------------------------------------------------------
def load_input_scene_image(ImageFile, ImType='Image', onvif_ip="", onvif_port=0, onvif_user="", onvif_passwd=""):
    """ Load input scene image (exit if fails) """

    if ImageFile == "onvif":
        mycam = onvif_camera(onvif_ip, onvif_port, onvif_user, onvif_passwd)
        imgOriginalScene = mycam.snaphot_capture()
    else:
        imgOriginalScene = imread(ImageFile)

    if imgOriginalScene is None:
        error("%s not read from file" % ImType)
        return None

    imgH, imgW, _ = imgOriginalScene.shape

    info("%s size: WxH=%dx%d" % (ImType, imgW, imgH))

    return imgOriginalScene

# ---------------------------------------------------------------------------------------------------------------
def isTemplateFound(imgOriginal, template, template_name, template_thr, debugMode):
    """ Seeking for police symbol in a multi-Scale template-matching fashion """

    # Convert original image to gray, and move to edge representation (boot + more accurate TM):
    gray = cvtColor(imgOriginal, COLOR_BGR2GRAY)
    edged_gray = Canny(gray, 50, 200)
    clone = dstack([edged_gray, edged_gray, edged_gray])

    # Convert template to gray:
    template_gray = cvtColor(template, COLOR_BGR2GRAY)

    # Loop over the scales of the template:
    found = None
    for scale in linspace(0.15, 1.0, 20)[::-1]:

        # Resize the template according to the scale, and keep track of the ratio of the resizing:
        resized = resize(template_gray, (0,0), fx=scale, fy=scale)
        (tH, tW) = resized.shape[:2]

        # Detect edges in the resized template, and apply template matching to find it in the edged image:
        edged_template = Canny(resized, 50, 200)
        result = matchTemplate(edged_gray, edged_template, TM_CCORR_NORMED)
        (_, maxVal, _, maxLoc) = minMaxLoc(result)

        # Check to see if the iteration should be visualized:
        if debugMode:

            # Draw a bounding box around the detected region
            intRandomBlue = randint(0, 255)
            intRandomGreen = randint(0, 255)
            intRandomRed = randint(0, 255)
            randColor = (intRandomBlue, intRandomGreen, intRandomRed)
            putText(clone, "maxVal=%.2f" % maxVal, (maxLoc[0], maxLoc[1]-20), FONT_HERSHEY_SIMPLEX, 1, randColor, 2)
            rectangle(clone, (maxLoc[0], maxLoc[1]), (maxLoc[0] + tW, maxLoc[1] + tH), randColor, 2)
            debug("maxVal=%.3f, tH=%d, tW=%d" % (maxVal, tH, tW), True)

        # If we have found a new maximum correlation value, then update the book-keeping variable:
        if found is None or maxVal > found[0]:
            found = (maxVal, maxLoc, tH, tW)

    if debugMode:
        imwrite("TemplateMatching_%s.jpg" % template_name, clone)

    if found is None:

        is_found = False
        loc = (None, None)

    else:

        # Unpack the book-keeping varaible and compute the (x, y) coordinates of the bounding box based on the resized ratio:
        (maxVal, maxLoc, tH, tW) = found
        (startX, startY) = (int(maxLoc[0]), int(maxLoc[1]))
        (endX, endY) = (int((maxLoc[0] + tW)), int((maxLoc[1] + tH)))
        loc = (maxLoc[0]+tW/2, maxLoc[1]+tH/2)
        avg_color = imgOriginal[startY:startY+tH, startX:startX+tW, :].mean()

        # draw a bounding box around the detected result and display the image
        is_found = maxVal >= template_thr[0] and avg_color > template_thr[1]

        if is_found:

            info("%s vehicle detected! (val1=%.2f, val2=%.2f)" % (template_name, found[0], avg_color))

            if debugMode:

                imgOriginalCopy = imgOriginal.copy()
                rectangle(imgOriginalCopy, (startX, startY), (endX, endY), (0, 255, 0), 2)
                putText(imgOriginalCopy, "maxVal=%.2f, avgRed=%.2f" % (maxVal, avg_color), (10, 30), FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
                imwrite("img_%s_symbol.jpg" % template_name, imgOriginalCopy)

    return (is_found, loc)

# ---------------------------------------------------------------------------------------------------------------
def sample_image(imgGray, sampleRect, maxVal):
    """ Sample image in a given rectangle, represented as [x0,y0,w,h], where x0,y0 is the top-left coordinates of the sampling region """

    y0, x0, w, h = sampleRect

    sampleSum = 0
    for y in range(h):
        for x in range(w):
            sampleSum += imgGray[y0+y][x0+x]

    sample = float(sampleSum) / (w * h * maxVal)

    return sample

File: Code/Cpp/Auxiliary.cpp
// Auxiliary.cpp
//  _     _                           ____  _       _         ____                            _ _   _
// | |   (_) ___ ___ _ __  ___  ___  |  _ \| | __ _| |_ ___  |  _ \ ___  ___ ___   __ _ _ __ (_) |_(_) ___  _ __
// | |   | |/ __/ _ \ '_ \/ __|/ _ \ | |_) | |/ _` | __/ _ \ | |_) / _ \/ __/ _ \ / _` | '_ \| | __| |/ _ \| '_ \
// | |___| | (__  __/ | | \__ \  __/ |  __/| | (_| | |_  __/ |  _ <  __/ (__ (_) | (_| | | | | | |_| | (_) | | | |
// |_____|_|\___\___|_| |_|___/\___| |_|   |_|\__,_|\__\___| |_| \_\___|\___\___/ \__, |_| |_|_|\__|_|\___/|_| |_|
//                                                                                |___/
// (c) Shahar Gino, July-2017, sgino209@gmail.com

#include "Auxiliary.hpp"

// ------------------------------------------------------------------------------------------------------------------------------
void generic_message(std::string message, std::string severity) {
    
    char buffer[100];
    sprintf(buffer, "%s: %s", severity.c_str(), message.c_str());
    
    std::cout << buffer << std::endl;
}

void info(std::string message)  { generic_message(message, "INFO");  }
void debug(std::string message) { generic_message(message, "DEBUG"); }
void error(std::string message) { generic_message(message, "ERROR"); }

// ------------------------------------------------------------------------------------------------------------------------------
void drawRedRectangleAroundPlate(cv::Mat &imgOriginalScene, PossiblePlate &licPlate) {
    
    cv::Point2f p2fRectPoints[4];
    
    // Get 4 vertices of rotated rectangle:
    licPlate.rrLocationOfPlateInScene.points(p2fRectPoints);
    
    // Draw 4 red lines:
    for (int i = 0; i < 4; i++) {
        cv::line(imgOriginalScene, p2fRectPoints[i], p2fRectPoints[(i + 1) % 4], SCALAR_RED, 2);
    }
}

// ------------------------------------------------------------------------------------------------------------------------------
void writeLicensePlateCharsOnImage(cv::Mat &imgOriginalScene, PossiblePlate &licPlate) {
    
    // Retrieve basic geometric metrics:
    cv::Point ptCenterOfTextArea;         // the center of the area the text will be written to
    cv::Point ptLowerLeftTextOrigin;      // the bottom left of the area that the text will be written to
    
    // Font styling (type, scale, thickness):
    int intFontFace = CV_FONT_HERSHEY_SIMPLEX;
    double dblFontScale = (double)licPlate.imgPlate.rows / 30.0;
    int intFontThickness = (int)round(dblFontScale * 1.5);
    int intBaseline = 0;
    cv::Size textSize = cv::getTextSize(licPlate.strChars, intFontFace, dblFontScale, intFontThickness, &intBaseline);
    
    // The horizontal location of the text area is the same as the plate:
    ptCenterOfTextArea.x = (int)licPlate.rrLocationOfPlateInScene.center.x;
    
    // The vertical location of the test area depends on the license-plate location in the scene image::
    if (licPlate.rrLocationOfPlateInScene.center.y < (imgOriginalScene.rows * 0.75)) {
        ptCenterOfTextArea.y = (int)round(licPlate.rrLocationOfPlateInScene.center.y) + (int)round((double)licPlate.imgPlate.rows * 1.6);  // chars below the plate
    }
    else {
        ptCenterOfTextArea.y = (int)round(licPlate.rrLocationOfPlateInScene.center.y) - (int)round((double)licPlate.imgPlate.rows * 1.6);  // chars above the plate
    }
    
    // Calculate the lower left origin of the text area based on the text area center, width, and height:
    ptLowerLeftTextOrigin.x = (int)(ptCenterOfTextArea.x - (textSize.width / 2));
    ptLowerLeftTextOrigin.y = (int)(ptCenterOfTextArea.y + (textSize.height / 2));
    
    // Write the text on the image:
    cv::putText(imgOriginalScene, licPlate.strChars, ptLowerLeftTextOrigin, intFontFace, dblFontScale, SCALAR_YELLOW, intFontThickness);
}

// ------------------------------------------------------------------------------------------------------------------------------
bool ROI_adjust(uint_x4_t ROI, cv::Mat &imgOriginalScene, int* W, int* H, uint_x4_t *adjROI) {
    
    cv::Size s = imgOriginalScene.size();
    int imgH = s.height;
    int imgW = s.width;
    bool autoRoiMode = false;
    
    if ((ROI.get_len() == 2) && (ROI.get_x1() == -1) && (ROI.get_x2() == -1)) {
        
        (*adjROI).set(int(imgW * W[0]/100.0), 1);
        (*adjROI).set(int(imgH * H[0]/100.0), 2);
        (*adjROI).set(int(imgW * W[1]/100.0), 3);
        (*adjROI).set(int(imgH * H[1]/100.0), 4);
        
        autoRoiMode = true;
        
        if (((W[0]+W[1]+W[2]) != 100) || ((H[0]+H[1]+H[2]) != 100)) {
            error("Invalid autoROI settings, doesnt converge to 100 percent");
        }
    }
    
    else if ((ROI.get_len() == 2) && (ROI.get_x1() == 0) && (ROI.get_x2() == 0)) {
        
        (*adjROI).set(0, 1);
        (*adjROI).set(0, 2);
        (*adjROI).set(imgW, 3);
        (*adjROI).set(imgH, 4);
    }
    
    else {
        *adjROI = ROI;
    }
    
    return autoRoiMode;
}

// ------------------------------------------------------------------------------------------------------------------------------
cv::Mat crop_roi_from_image(cv::Mat &imgOriginalScene, uint_x4_t ROI, bool autoRoiMode) {
    
    int imgW;
    int imgH;
    unsigned int roiW;
    unsigned int roiH;
    char buffer[100];
    
    if (ROI.get_x3() == -1 || ROI.get_x4() == -1) {
        
        cv::Size s = imgOriginalScene.size();
        imgH = s.height;
        imgW = s.width;
        roiW = imgW;
        roiH = imgH;
    }
    else {
        
        roiW = ROI.get_x3();
        roiH = ROI.get_x4();
    }
    
    double roiCx = ROI.get_x1() + roiW / 2.0;
    double roiCy = ROI.get_x2() + roiH / 2.0;
    
    cv::Mat imgCropped;
    cv::getRectSubPix(imgOriginalScene, cv::Size2f((float)roiW, (float)roiH), cv::Point2f((float)roiCx, (float)roiCy), imgCropped);
    
    sprintf(buffer, "ROI size: (Cx,Cy)=(%.2f,%.2f), WxH=%dx%d, autoROI=%d", roiCx, roiCy, roiW, roiH, int(autoRoiMode));
    info(buffer);
    
    return imgCropped;
}

// ------------------------------------------------------------------------------------------------------------------------------
cv::Mat load_input_scene_image(std::string imageFile, std::string ImType,
                               std::string onvif_ip, unsigned int onvif_port, std::string onvif_user, std::string onvif_passwd) {
    
    cv::Mat imgOriginalScene;
    char buffer[100];
    
    if (imageFile == "onvif") {
        onvif_camera mycam = onvif_camera(onvif_ip, onvif_port, onvif_user, onvif_passwd);
        imgOriginalScene = mycam.snaphot_capture();
    } else {
        imgOriginalScene = cv::imread(imageFile);
    }
    
    if (imgOriginalScene.empty()) {
        sprintf(buffer, "%s not read from file", ImType.c_str());
        error(buffer);
        #if defined(WIN32) || defined(_WIN32) || defined(__WIN32) && !defined(__CYGWIN__)
        _getch();
        #else
        wgetch(stdscr);
        #endif
    }
    
    cv::Size s = imgOriginalScene.size();
    int imgH = s.height;
    int imgW = s.width;
    sprintf(buffer, "%s size: WxH=%dx%d", ImType.c_str(), imgW, imgH);
    info(buffer);
    
    return imgOriginalScene;
}

// ------------------------------------------------------------------------------------------------------------------------------
bool isTemplateFound(cv::Mat &imgOriginal, cv::Mat &imgTemplate, std::string template_name, double_xy_t template_thr, double_xy_t *templateLoc, bool debugMode) {

    cv::RNG rng;
    bool is_found = false;
    char buffer[100];
    double minVal, maxVal;
    int result_cols, result_rows, tH=0, tW=0, startX, startY, endX, endY;
    cv::Point minLoc, maxLoc;
    cv::Mat gray, edged_gray, template_gray, result, edged_template;
    
    // Convert original image to gray, and move to edge representation (boot + more accurate TM):
    cv::cvtColor(imgOriginal, gray, CV_BGR2GRAY);
    Canny(gray, edged_gray, 50, 200);
    cv::Mat clone;
    cv::cvtColor(edged_gray, clone, CV_GRAY2RGB);

    // Convert template to gray:
    cv::cvtColor(imgTemplate, template_gray, CV_BGR2GRAY);

    // Loop over the scales of the template:
    double found[5] = {0, 0, 0, 0, 0};
    std::vector<double> scales = linspace(0.15, 1.0, 20);
    std::vector<double>::reverse_iterator scale = scales.rbegin();
    for (; scale!= scales.rend(); ++scale) {

        // Resize the template according to the scale, and keep track of the ratio of the resizing:
        
        cv::Mat template_gray_cloned = template_gray.clone();
        cv::Mat resized;
        cv::resize(template_gray_cloned, resized, cv::Size(), *scale, *scale);
        tH = resized.rows;
        tW = resized.cols;
        
        // Detect edges in the resized template, and apply template matching to find it in the edged image:
        Canny(resized, edged_template, 50, 200);
        result_cols =  edged_gray.cols - edged_template.cols + 1;
        result_rows = edged_gray.rows - edged_template.rows + 1;
        result.create( result_rows, result_cols, CV_32FC1 );
        cv::matchTemplate(edged_gray, edged_template, result, cv::TM_CCORR_NORMED);
        minMaxLoc( result, &minVal, &maxVal, &minLoc, &maxLoc, cv::Mat() );

        // Check to see if the iteration should be visualized:
        if (debugMode) {

            // Draw a bounding box around the detected region
            int intRandomBlue = rng.uniform(0, 256);
            int intRandomGreen = rng.uniform(0, 256);
            int intRandomRed = rng.uniform(0, 256);
            cv::Scalar randColor = cv::Scalar((double)intRandomBlue, (double)intRandomGreen, (double)intRandomRed);
            sprintf(buffer, "maxVal=%.2f", maxVal);
            cv::putText(clone, buffer, cv::Point(maxLoc.x, maxLoc.y-20), cv::FONT_HERSHEY_SIMPLEX, 1, randColor, 2);
            cv::rectangle(clone, cv::Rect(maxLoc.x, maxLoc.y, tW, tH), randColor, 2);
            sprintf(buffer, "maxVal=%.3f, tH=%d, tW=%d", maxVal, tH, tW);
            debug(buffer);
        }

        // If we have found a new maximum correlation value, then update the book-keeping variable:
        if (maxVal > found[0]) {
            found[0] = maxVal;
            found[1] = maxLoc.x;
            found[2] = maxLoc.y;
            found[3] = tH;
            found[4] = tW;
        }
    }
    
    if (debugMode) {
        sprintf(buffer, "TemplateMatching_%s.jpg", template_name.c_str());
        cv::imwrite(buffer, clone);
    }
    
    if (templateLoc && (found[0] == 0)) {

        *templateLoc = double_xy_t(0, 0);
    }
    
    else {

        // Unpack the book-keeping varaible and compute the (x, y) coordinates of the bounding box based on the resized ratio:
        maxVal = found[0];
        maxLoc = cv::Point(found[1],found[2]);
        tH = int(found[3]);
        tW = int(found[4]);
        startX = int(maxLoc.x);
        startY = int(maxLoc.y);
        endX = int((maxLoc.x + tW));
        endY = int((maxLoc.y + tH));
        
        cv::Mat imgCropped;
        cv::getRectSubPix(imgOriginal, cv::Size2f((float)tW, (float)tH), cv::Point2f((float)(maxLoc.x+tW/2), (float)(maxLoc.y+tH/2)), imgCropped);
        cv::Scalar tempVal = mean(imgCropped);
        double avg_color = (tempVal.val[0] + tempVal.val[1] + tempVal.val[2]) / 3;

        // draw a bounding box around the detected result and display the image
        is_found = (maxVal >= template_thr.get_x()) && (avg_color > template_thr.get_y());

        if (is_found) {

            sprintf(buffer, "%s vehicle detected! (val1=%.2f, val2=%.2f)", template_name.c_str(), found[0], avg_color);
            info(buffer);

            if (debugMode) {

                cv::Mat imgOriginalCopy = imgOriginal.clone();
                cv::rectangle(imgOriginalCopy, cv::Rect(startX, startY, endX-startX, endY-startY), SCALAR_GREEN, 2);
                sprintf(buffer, "maxVal=%.2f, avgRed=%.2f", maxVal, avg_color);
                cv::putText(imgOriginalCopy, buffer, cv::Point(10,30), cv::FONT_HERSHEY_SIMPLEX, 1, SCALAR_GREEN, 2);
                sprintf(buffer, "img_%s_symbol.jpg", template_name.c_str());
                cv::imwrite(buffer, imgOriginalCopy);
            }
        }
    }

    if (templateLoc) {
        *templateLoc = double_xy_t(maxLoc.x+tW/2, maxLoc.y+tH/2);
    }
    
    return is_found;
}

// ------------------------------------------------------------------------------------------------------------------------------
std::vector<double> linspace(double min, double max, int n)
{
    std::vector<double> result;
    // vector iterator
    int iterator = 0;
    
    for (int i = 0; i <= n-2; i++)
    {
        double temp = min + i*(max-min)/(floor((double)n) - 1);
        result.insert(result.begin() + iterator, temp);
        iterator += 1;
    }
    
    result.insert(result.begin() + iterator, max);
    return result;
}

// ------------------------------------------------------------------------------------------------------------------------------
bool is_digits(const std::string &str)
{
    return str.find_first_not_of("0123456789") == std::string::npos;
}

// ------------------------------------------------------------------------------------------------------------------------------
double sample_image(cv::Mat &imgGray, uint_x4_t sampleRect, unsigned int maxVal) {
    
    unsigned int y0 = sampleRect.get_x1();
    unsigned int x0 = sampleRect.get_x2();
    unsigned int w  = sampleRect.get_x3();
    unsigned int h  = sampleRect.get_x4();
    
    unsigned int sampleSum = 0;
    for (unsigned int y=0; y<h; y++) {
        for (unsigned int x=0; x<w; x++) {
            sampleSum += imgGray.at<unsigned char>(y0+y, x0+x);
        }
    }
    
    double sample = double(sampleSum) / (w * h * maxVal);
    
    return sample;
}
